src/github.com/matrix-org/dendrite/appservice/workers/transaction_scheduler.go

```diff
@@ -0,0 +1,239 @@
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"math"
+	"net/http"
+	"time"
+
+	"github.com/matrix-org/dendrite/appservice/storage"
+	"github.com/matrix-org/dendrite/appservice/types"
+	"github.com/matrix-org/dendrite/common/config"
+	"github.com/matrix-org/gomatrixserverlib"
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	// Maximum size of events sent in each transaction.
+	// Warning, if this is lowered and a number of events greater than the previous
+	// batch size were still to be sent, then a number of events equal to the
+	// difference will be ignored by the app service.
+	// TL;DR: Don't lower this number with any AS events still left in the database.
+	transactionBatchSize = 50
+	// Timeout for sending a single transaction to an application service.
+	transactionTimeout = time.Second * 60
+	// The current transaction ID. Increments after every successful transaction.
+	currentTransactionID = 0
+)
+
+// SetupTransactionWorkers spawns a separate goroutine for each application
+// service. Each of these "workers" handle taking all events intended for their
+// app service, batch them up into a single transaction (up to a max transaction
+// size), then send that off to the AS's /transactions/{txnID} endpoint. It also
+// handles exponentially backing off in case the AS isn't currently available.
+func SetupTransactionWorkers(
+	appserviceDB *storage.Database,
+	workerStates []types.ApplicationServiceWorkerState,
+) error {
+	// Create a worker that handles transmitting events to a single homeserver
+	for _, workerState := range workerStates {
+		// Don't create a worker if this AS doesn't want to receive events
+		if workerState.AppService.URL != "" {
+			go worker(appserviceDB, workerState)
+		}
+	}
+	return nil
+}
+
+// worker is a goroutine that sends any queued events to the application service
+// it is given.
+func worker(db *storage.Database, ws types.ApplicationServiceWorkerState) {
+	log.Infof("Starting Application Service %s", ws.AppService.ID)
+	ctx := context.Background()
+
+	// Initialize transaction ID counter
+	var err error
+	currentTransactionID, err = db.GetTxnIDWithAppServiceID(ctx, ws.AppService.ID)
+	if err != nil && err != sql.ErrNoRows {
+		log.WithError(err).Fatalf("appservice %s worker unable to get latest transaction ID from DB",
+			ws.AppService.ID)
+		return
+	}
+
+	// Grab the HTTP client for sending requests to app services
+	client := &http.Client{
+		Timeout: transactionTimeout,
+	}
+
+	// Initial check for any leftover events to send from last time
+	eventCount, err := db.CountEventsWithAppServiceID(ctx, ws.AppService.ID)
+	if err != nil {
+		log.WithError(err).Fatalf("appservice %s worker unable to read queued events from DB",
+			ws.AppService.ID)
+		return
+	}
+	ws.Cond.L.Lock()
+	*ws.EventsReady = eventCount
+	ws.Cond.L.Unlock()
+
+	// Loop forever and keep waiting for more events to send
+	for {
+		// Wait for more events if we've sent all the events in the database
+		if *ws.EventsReady <= 0 {
```

Comment:
This check should happen in the lock. I think things can spuriously wake up (they can in most other langs), so may people tend to do:

```golang
ws.Cond.L.Lock()
for *ws.EventsReady <= 0 {
    ws.Cond.Wait()
}
ws.Cond.L.Unlock()
```